### PR TITLE
contracts-bedrock: improve `CrossL2Inbox` devex

### LINKED ISSUE #11365: Interop: Add `validateMessage` to `CrossL2Inbox`
# Overview

Add a new function to the `CrossL2Inbox` called `validateMessage(Identifier,bytes32)` that emits the `ExecutingMessage` event. This is a UX improvement for the `CrossL2Inbox`.

# Definition of Done

The function is implemented and has full test coverage.

### PR DESCRIPTION
**Pull request by tynes**
**Description**

Improve the `CrossL2Inbox` devex by creating an alternative entrypoint.
This design was not considered previously because there was a "top level
call" restriction, aka the "only EOA" invariant. This was to allow for
static analysis of transactions, keeping resource usage lower for
validating transactions when building blocks or at the mempool layer.
Since 3074/7702 render the enforcement of only eoa impossible,  we decided
to op/acc and lean into the approach of allowing subcalls to trigger
`ExecutingMessage` events.

This new interface allows another contract to be the entrypoint,
the idea is that the user sends the `Identifier` and the serialized
log (message) to whatever contract that they want and then pass it to
`CrossL2Inbox.validateMessage` which then emits the event that consensus
validates. This allows the calling smart contract to be aware of the
schema for the log and deserialize it however they see fit. Since the
serialized logs are done with the following algorithm:

```go
msg := make([]byte, 0)
for _, topic := range log.Topics {
    msg = append(msg, topic.Bytes()...)
}
msg = append(msg, log.Data...)
```

It is very easy to use `abi.decode` to decode a log, given that solidity
was used to `emit` it. The topics are `bytes32` and then the data is
abi encoded given the schema of the event itself. Unused parts like
`topic[0]` (hash of the event name) can be dropped when decoding if
they are not required.

### CODE REVIEW

File: packages/contracts-bedrock/src/L2/CrossL2Inbox.sol
Comment (skeletor-spaceman) [inline packages/contracts-bedrock/src/L2/CrossL2Inbox.sol]:
```suggestion
    function validateMessage(Identifier calldata _id, bytes32 _msgHash) external nonReentrant {
        // Check the Identifier.
        _checkIdentifier(_id);

        emit ExecutingMessage(_msgHash, _id);
    }
```
since we don't really need the entire message for executing :)

Comment (tynes) [reply packages/contracts-bedrock/src/L2/CrossL2Inbox.sol]: Ah good catch

Comment (tynes) [reply packages/contracts-bedrock/src/L2/CrossL2Inbox.sol]: This is a very helpful optimization that could save lots of gas over the long run

**Comment by tynes** [comment]
This needs to be specified before we can consider merging it

File: packages/contracts-bedrock/src/libraries/TransientContext.sol
Comment (tynes) [inline packages/contracts-bedrock/src/libraries/TransientContext.sol]: we will want to update the key here to be more generic so this pattern can be used in any contract

File: packages/contracts-bedrock/src/libraries/TransientContext.sol
Comment (tynes) [inline packages/contracts-bedrock/src/libraries/TransientContext.sol]: This could all be assembly tbh

**Comment by fainashalts** [comment]
I love this idea! DevX team spoke about it today, and I'm wondering if this user story from @jakim929 would be covered by this change: I want to update the contract state on chain B anytime OpenseaContract#OrderFulfilled is emitted on Chain A (until Opensea actually upgrades their contract to be "interop aware").

**Comment by skeletor-spaceman** [comment]
> I want to update the contract state on chain B anytime OpenseaContract#OrderFulfilled is emitted on Chain A (until Opensea actually upgrades their contract to be "interop aware").

exactly!

**Comment by tynes** [comment]
There are 2 pieces of work happening in this PR:
- Moves the transient reentrency lock out of the `L2ToL2CrossDomainMessenger` into the `TransientReentrancyAware` contract. This enables the reentrency lock logic to be reused. We may want to standardize on how the implementations work a bit
- Adds a new function to the `CrossL2Inbox` which allows for better developer experience. This function needs testing and it also needs to be added to the spec

File: packages/contracts-bedrock/src/libraries/TransientContext.sol
Comment (AmadiMichael) [reply packages/contracts-bedrock/src/libraries/TransientContext.sol]: this good `bytes32(uint256(keccak256("transientreentrancyaware.entered")) - 1)`, or too long?

File: packages/contracts-bedrock/src/L2/CrossL2Inbox.sol
Comment (tynes) [inline packages/contracts-bedrock/src/L2/CrossL2Inbox.sol:142]: If you add some natspec here then this is good to merge

File: packages/contracts-bedrock/src/libraries/TransientContext.sol
Comment (tynes) [reply packages/contracts-bedrock/src/libraries/TransientContext.sol]: this is good! it just needs to be unique so there is no collision with something that uses this contract

File: packages/contracts-bedrock/semver-lock.json
Comment (tynes) [inline packages/contracts-bedrock/semver-lock.json:70]: You can see in the `semver-lock.json` that both the `CrossL2Inbox` and the `L2ToL2CrossDomainMessenger` had their sourcecode + bytecode modified, so both should receive a version bump. Since they are prerelease, its good to bump the right most integer

**Comment by tynes** [comment]
Looks like the `contracts-bedrock-checks` are failing - right now you need to run `pnpm snapshots` locally and commit the results of that. The `kontrol` artifacts are the output and they are used for formal verification jobs in CI

File: packages/contracts-bedrock/test/L2/CrossL2Inbox.t.sol
Comment (mds1) [inline packages/contracts-bedrock/test/L2/CrossL2Inbox.t.sol]: For `expectEmit` let's default to using all `true`, which this suggestion is equivalent to. This is preferred as it's a stronger test, and if we every change topics or parameters we ensure everything is tested by default
```suggestion
        vm.expectEmit(Predeploys.CROSS_L2_INBOX);
```

Comment (AmadiMichael) [reply packages/contracts-bedrock/test/L2/CrossL2Inbox.t.sol]: noted, done.

File: packages/contracts-bedrock/test/L2/CrossL2Inbox.t.sol
Comment (tynes) [inline packages/contracts-bedrock/test/L2/CrossL2Inbox.t.sol]: small nit: just using `vm.expectEmit(address)` is the same as `vm.expectEmit(true,true,true,true,address(0x..)`. We prefer the first because it is less code

**Comment by tynes** [comment]
I cannot approve this as I opened the original PR but it looks good to me

**Review by skeletor-spaceman — APPROVED** [review]

**Comment by tynes** [comment]
Failing fuzz test:

```
Failing tests:
Encountered 4 failing tests in test/libraries/SafeCall.t.sol:SafeCall_Test
[FAIL. Reason: call not successful; counterexample: calldata=0x060874140000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000a00000000000000000000000000000000000000000000000000000000000021390000000000000000000000000000000000000000000000000000000000000c6900000000000000000000000000000000000000000000000000000000000000a000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000044d7f0a5 args=[0x0000000000000000000000000000000000000000, 0x000000000000000000000000000000000000000A, 8505, 3177, 0x0000000000000000000000000000000000000000000000000000000044d7f0a5]] testFuzz_callWithMinGas_hasEnough_succeeds(address,address,uint64,uint64,bytes) (runs: 301, μ: 54299, ~: 55304)
[FAIL. Reason: call not successful; counterexample: calldata=0xe28f97dc0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000a00000000000000000000000000000000000000000000000000000000000021390000000000000000000000000000000000000000000000000000000000000c6900000000000000000000000000000000000000000000000000000000000000a000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000044d7f0a5 args=[0x0000000000000000000000000000000000000000, 0x000000000000000000000000000000000000000A, 8505, 3177, 0x0000000000000000000000000000000000000000000000000000000044d7f0a5]] testFuzz_call_succeeds(address,address,uint256,uint64,bytes) (runs: 0, μ: 0, ~: 0)
[FAIL. Reason: send not successful; counterexample: calldata=0x491205050000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000a00000000000000000000000000000000000000000000000000000000000021390000000000000000000000000000000000000000000000000000000000000c69 args=[0x0000000000000000000000000000000000000000, 0x000000000000000000000000000000000000000A, 8505, 3177]] testFuzz_sendWithGas_succeeds(address,address,uint64,uint256) (runs: 0, μ: 0, ~: 0)
[FAIL. Reason: send not successful; counterexample: calldata=0x33717fe30000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000a0000000000000000000000000000000000000000000000000000000000002139 args=[0x0000000000000000000000000000000000000000, 0x000000000000000000000000000000000000000A, 8505]] testFuzz_send_succeeds(address,address,uint256) (runs: 0, μ: 0, ~: 0)
```

**Pull request merged**
